src/event_handlers/hubV2.ts

```diff
@@ -367,19 +400,64 @@ HubV2.TransferSingle.handlerWithLoader({
   },
 });
 
-HubV2.TransferBatch.handler(
-  async ({ event, context }) =>
-    await handleTransfer({
+HubV2.TransferBatch.handlerWithLoader({
+  loader: async ({ event, context }) => {
+    const tokenIds = event.params.ids.map((id) => id.toString());
+
+    // flatten all promises into a single array
+    const allPromises = [
+      ...tokenIds.map((id) => context.Token.get(id)),
+      ...tokenIds.flatMap((tokenId) => [
+        context.AvatarBalance.get(
+          makeAvatarBalanceEntityId(event.params.to, tokenId)
+        ),
+        context.AvatarBalance.get(
+          makeAvatarBalanceEntityId(event.params.from, tokenId)
+        ),
+      ]),
+    ];
+
+    const allResults = await Promise.all(allPromises);
+
+    const tokens = allResults.slice(0, tokenIds.length) as (
+      | Token_t
+      | { id: string }
+    )[];
+    const allAvatarBalances = allResults.slice(tokenIds.length) as (
+      | AvatarBalance_t
+      | { avatar_id: string; token_id: string }
+    )[];
+
+    const avatarsBalance = tokenIds.map((tokenId, i) => ({
+      to: allAvatarBalances[i * 2] ?? {
+        avatar_id: event.params.to,
+        token_id: tokenId,
+      },
+      from: allAvatarBalances[i * 2 + 1] ?? {
```

Comment:
what is this sorcery? 😂  why i*2 ?

Reply:
because i needs AvatarBalance "from" user and "to" user. Given that they are all in one single array, then it jumps each two entities
0-1
2-3
...